plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/aggregation/AggregateSplitSource.java

```diff
@@ -0,0 +1,485 @@
+            if (requiresColumnStats) {
+                scan = scan.includeColumnStats();
+            }
+            this.fileScanTaskIterable = TableScanUtil.splitFiles(scan.planFiles(), tableScan.targetSplitSize());
+            closer.register(fileScanTaskIterable);
+            this.fileScanTaskIterator = fileScanTaskIterable.iterator();
+            closer.register(fileScanTaskIterator);
+            // TODO: Remove when NPE check has been released: https://github.com/trinodb/trino/issues/15372
+            isFinished();
+        }
+
+        TupleDomain<IcebergColumnHandle> dynamicFilterPredicate = dynamicFilter.getCurrentPredicate()
+                .transformKeys(IcebergColumnHandle.class::cast);
+        if (dynamicFilterPredicate.isNone()) {
+            finish();
+            return completedFuture(NO_MORE_SPLITS_BATCH);
+        }
+
+        // Note: not using the maxSize of the batch, as we need to calculate the aggregates in single threaded.
+        // When we will implement it in distributed way which might need the Engine to support Top level aggregation.
+        // As right now if connector handles aggregation pushdown, Engine does not handle the aggregate function.
+        ImmutableList.Builder<ConnectorSplit> splits = ImmutableList.builder();
+        while (fileScanTaskIterator.hasNext()) {
+            FileScanTask scanTask = fileScanTaskIterator.next();
+
+            if (!uniqueFileName.add(scanTask.file().path().toString())) {
+                // duplicate file entry, mostly because of the file size and Iceberg's internal default file-size=128 MB
+                // so do not need to process this entry, as this data file's metadata is already being used.
+                continue;
+            }
+
+            if (scanTask.deletes().isEmpty() &&
+                    maxScannedFileSizeInBytes.isPresent() &&
+                    scanTask.file().fileSizeInBytes() > maxScannedFileSizeInBytes.get()) {
+                continue;
+            }
+
+            if (!pathDomain.includesNullableValue(utf8Slice(scanTask.file().path().toString()))) {
+                continue;
+            }
+            if (!fileModifiedTimeDomain.isAll()) {
+                long fileModifiedTime = getModificationTime(scanTask.file().path().toString());
+                if (!fileModifiedTimeDomain.includesNullableValue(packDateTimeWithZone(fileModifiedTime, UTC_KEY))) {
+                    continue;
+                }
+            }
+
+            Schema fileSchema = scanTask.spec().schema();
+            Map<Integer, Optional<String>> partitionKeys = getPartitionKeys(scanTask);
+
+            Set<IcebergColumnHandle> identityPartitionColumns = partitionKeys.keySet().stream()
+                    .map(fieldId -> getColumnHandle(fileSchema.findField(fieldId), typeManager))
+                    .collect(toImmutableSet());
+
+            Supplier<Map<ColumnHandle, NullableValue>> partitionValues = memoize(() -> {
```

Comment:
The way how `AggregateSplitSource` duplicates all the code from `IcebergSplitSource` is not maintainable.
I guess it was already witnessed during rebasing the PR. The copy needs manual updates to keep in sync.

I think instead it would be better to have aggregate split source act as a decorator / post-processor for `IcebergSplitSource`.

Reply:
sure, @findepi 
yeah, I had mentioned that keeping the dupe for now, to get a buy in on the overall design and then can change/refactor as well.
https://github.com/trinodb/trino/pull/15832#discussion_r1086165192